### PR TITLE
APPLE-79 Refinements to the slug component

### PR DESCRIPTION
**Pull request by kevinfodness**
* Add a postmeta entry for `apple_news_slug` so that it can be set in the admin instead of only via a filter.
* Add a field to the Gutenberg interface and the metabox for setting the slug value.
* Refactor metabox save to loop over field definitions to reduce code duplication.
* Standardize naming convention from `onUpdate*` to `onChange*` to match Gutenberg conventions.
* Fix order of slug settings in theme JSON files.
* Add setting for slug dark mode color.
* Refactor component tests as end-to-end tests using the `get_json_for_post` function in the `Apple_News_Testcase` base class.

### CODE REVIEW

File: admin/class-admin-apple-meta-boxes.php [hunk 1]
```diff
@@ -167,6 +167,36 @@ public static function save_post_meta( $post_id ) {
 		// Check the nonce.
 		check_admin_referer( self::PUBLISH_ACTION, 'apple_news_nonce' );
 
+		// Save straightforward fields.
+		$fields = [
```
Comment (jomurgel) [inline admin/class-admin-apple-meta-boxes.php:171]: So much cleaner! :chef-kiss:

File: tests/apple-exporter/components/test-class-cover.php [hunk 2]
```diff
@@ -6,289 +6,125 @@
...
-	public function testCaptionSetting() {
-		// Create an image and give it a caption.
-		$image_id = $this->get_new_attachment( 0, 'Test Caption!' );
-
-		// Create a test post.
-		$post_id = self::factory()->post->create();
-
-		// Set the featured image for the post.
-		set_post_thumbnail( $post_id, $image_id );
+	public function test_render() {
```
Comment (jomurgel) [inline tests/apple-exporter/components/test-class-cover.php:77]: My failure is being able to understand WP PHPUnit. I'd love to write unit tests for my PHP functions, but damnitall if this makes no reasonable sense to me.

**Review by jomurgel — APPROVED** [review]
Looks good. Got a little hung up on the unit tests, but I think I follow and assume they return without errors. 🍞

File: tests/apple-exporter/components/test-class-cover.php (hunk 2 above)
Comment (kevinfodness) [reply tests/apple-exporter/components/test-class-cover.php:77]: I've built some helper functions into Apple News to help with some of this, like the `get_json_for_post` function and the `set_theme_settings` function, and the `get_new_attachment` function. Those are defined in `Apple_News_Testcase`. The `self::factory()->post->create()` function is part of the WP extensions to PHPUnit. Basically, `self::factory()` returns an object that contains a few different factories for creating things, like posts, users, terms, etc. I'm using the post factory to create a new post. The arguments you pass to it are basically the same ones you would pass to `wp_insert_post`. The major difference is that the factory will insert sample content for you into all fields that you don't specify values for. Here, I'm specifying the post content, but not the excerpt, title, etc, so those will all have values, but they will be things like `Title 00001` or whatever.

Function calls like `set_post_thumbnail` and `add_post_meta` are normal WordPress functions that modify the state of the test database and set up data for tests. The WordPress PHPUnit test runner resets the database state between tests, so we can insert test data and query against it, knowing that it will get wiped out between runs.

The `$this->assertEquals` calls are stock PHPUnit and are testing our assumptions about what is being returned.

Basically, the old tests were unit tests that tested the behavior of a specific class. These new tests are end-to-end tests, because we're publishing a new post in the database and getting the JSON for it using the Publish to Apple News plugin, and we're verifying that the JSON it generates is what we expect for the class that is responsible for the piece of the JSON that we're testing.

I'd be happy to do a deeper dive into this with you, if you'd like.
